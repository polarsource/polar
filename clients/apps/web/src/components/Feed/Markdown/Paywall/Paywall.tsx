import { Article } from '@polar-sh/sdk'
import Link from 'next/link'
import { twMerge } from 'tailwind-merge'

const Paywall = (props: {
  children?: React.ReactNode
  showPaywalledContent: boolean
  isSubscriber: boolean
  article: Article
  renderer?: typeof BrowserPaywall
}) => {
  if (props.renderer) {
    const C = props.renderer
    return (
      <C
        showPaywalledContent={props.showPaywalledContent}
        isSubscriber={props.isSubscriber}
        article={props.article}
      >
        {props.children}
      </C>
    )
  }

  return (
    <BrowserPaywall
      showPaywalledContent={props.showPaywalledContent}
      isSubscriber={props.isSubscriber}
      article={props.article}
    >
      {props.children}
    </BrowserPaywall>
  )
}

const BasePaywall = (props: {
  showPaywalledContent: boolean
  isSubscriber: boolean
  article: Article
  children?: React.ReactNode
  classNames?: string
}) => {
  if (
    props.showPaywalledContent === false ||
    !props.children ||
    (Array.isArray(props.children) && props.children.length === 0)
  ) {
    return (
      <div
        className={twMerge(
          'my-4 flex flex-col items-center rounded-3xl bg-gray-100 px-8 py-4',
          props.classNames,
        )}
      >
        {props.isSubscriber ? (
          <p>
            This section is for premium subscribers only. Upgrade{' '}
            <Link
              className="text-blue-500 hover:text-blue-400 dark:text-blue-400 dark:hover:text-blue-300"
              href={`/${props.article.organization.name}/subscriptions`}
            >
              your subscription
            </Link>{' '}
            to a tier with the &quot;Paid Subscription&quot; benefit to get
            access to it.
          </p>
        ) : (
          <p>
            This section is for premium subscribers only. Subscribe to{' '}
            <Link
              className="text-blue-500 hover:text-blue-400 dark:text-blue-400 dark:hover:text-blue-300"
              href={`/${props.article.organization.name}/subscriptions`}
            >
              {props.article.organization.pretty_name ||
                props.article.organization.name}
            </Link>{' '}
            to get access to it.
          </p>
        )}
      </div>
    )
  }
  return <p>{props.children}</p>
}

const BrowserPaywall = (props: {
  showPaywalledContent: boolean
  isSubscriber: boolean
  article: Article
  children?: React.ReactNode
}) => {
  return <BasePaywall {...props} classNames="dark:bg-polar-700" />
}

export const EmailPaywall = (props: {
  showPaywalledContent: boolean
  isSubscriber: boolean
  article: Article
  children?: React.ReactNode
}) => {
  return <BasePaywall {...props} />
}

export default Paywall
