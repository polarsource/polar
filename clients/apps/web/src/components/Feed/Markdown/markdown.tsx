import { AdvertisementDisplay, Article } from '@polar-sh/sdk'
import React, { Fragment } from 'react'

export type RenderArticle = Pick<
  Article,
  | 'title'
  | 'body'
  | 'published_at'
  | 'byline'
  | 'organization'
  | 'slug'
  | 'is_preview'
>

export type BenefitAds = {
  benefitId: string
  ads: Array<AdvertisementDisplay>
}

// strictCreateElement removes unsupported types and attributes
export const wrapStrictCreateElement = (args: {
  article: RenderArticle
  showPaywalledContent?: boolean
  isSubscriber?: boolean
  paidArticlesBenefitName?: string
  defaultOverride?: React.FunctionComponent
  extraAllowedCustomComponents?: string[]
  adsContext?: BenefitAds[]
}): ((
  type:
    | string
    | import('react').FunctionComponent<{}>
    | import('react').ComponentClass<{}, any>,
  props: JSX.IntrinsicAttributes | any,
  children: any,
) => JSX.Element) => {
  const strictCreateElement = (
    type:
      | string
      | import('react').FunctionComponent<{}>
      | import('react').ComponentClass<{}, any>,
    props: JSX.IntrinsicAttributes | any,
    children: any,
  ): JSX.Element => {
    const allowedTypes = [
      'a',
      'b',
      'blockquote',
      'code',
      'del',
      'div',
      'em',
      'footer',
      'h1',
      'h2',
      'h3',
      'h4',
      'h5',
      'h6',
      'hr',
      'img',
      'input',
      'li',
      'ol',
      'p',
      'picture',
      'pre',
      'source',
      'span',
      'strong',
      'sub',
      'sup',
      'table',
      'tbody',
      'td',
      'th',
      'thead',
      'tr',
      'ul',
    ]

    // clean up props, only pass down a limited set of safe props
    let trimProps: typeof props = {
      key: props?.key,
      style: props?.style,
      align: props?.align,
      className: props?.className,
      id: props?.id,
    }

    const allowedCustomComponents = [
      // custom completely overridden components
      'embed',
      'iframe',
      'img',
      'pre',
      // our custom components
      'poll',
      'paywall',
      'subscribenow',
      ...(args.extraAllowedCustomComponents ?? []),
    ].map((s) => s.toLowerCase())

    // Custom components
    if (
      typeof type === 'function' &&
      type?.name &&
      allowedCustomComponents.includes(type?.name.toLowerCase())
    ) {
      const customComponentName = type?.name.toLowerCase()

      if (customComponentName === 'img') {
        trimProps.src = props?.src
        trimProps.height = props?.height
        trimProps.width = props?.width
        trimProps.alt = props?.alt
        children = undefined // can never have children
      }

      if (customComponentName === 'embed') {
        trimProps.src = props?.src
        children = undefined // can never have children
      }

      if (customComponentName === 'iframe') {
        trimProps.src = props?.src
        trimProps.title = props?.title
        trimProps.allow = props?.allow
        children = undefined // can never have children
      }

      if (['subscribenow', 'paywall'].includes(customComponentName)) {
        trimProps.article = args.article
        // Default to true in the client side renderer. When rendering posts for end-users, the premium content is already stripped out by the backend.
        trimProps.showPaywalledContent = args.showPaywalledContent ?? true
        trimProps.isSubscriber = args.isSubscriber ?? false
        trimProps.paidArticlesBenefitName = args.paidArticlesBenefitName
      }

      if (customComponentName === 'ad') {
        trimProps.subscriptionBenefitId = props?.subscriptionBenefitId
        trimProps.adsContext = args.adsContext
      }

      return React.createElement(
        type,
        {
          ...trimProps,
        } as JSX.IntrinsicAttributes,
        children,
      )
    }

    if (typeof type !== 'string') {
      return <></>
    }

    type = type.toLowerCase()

    if (!allowedTypes.includes(type)) {
      return <></>
    }

    if (args.defaultOverride) {
      return React.createElement(args.defaultOverride, {}, props.children)
    }

    if (type === 'a') {
      const href = props?.href
      if (
        typeof href === 'string' &&
        // Restricted protos.
        // Do not allow relative URLs.
        (href.startsWith('https://') ||
          href.startsWith('http://') ||
          href.startsWith('mailto://') ||
          href.startsWith('#'))
      ) {
        trimProps.href = props?.href
      } else {
        return <></>
      }
    }

    if (type === 'input') {
      if (props?.type !== 'checkbox') {
        return <></>
      }

      trimProps.type = 'checkbox'
      trimProps.checked = props?.checked
      trimProps.disabled = 'disabled'
      children = undefined // can never have children
    }

    if (type === 'source') {
      trimProps.media = props?.media
      trimProps.srcSet = props?.srcSet
    }

    if (type === 'p' && Array.isArray(children)) {
      return React.createElement(
        type,
        trimProps,
        <Fragment>
          {children.map((ch, idx) => {
            // Double whitespace as newline.
            if (typeof ch === 'object' && ch.key === null) {
              return (
                <Fragment key={idx}>
                  <br />
                </Fragment>
              )
            }

            return <Fragment key={idx}>{ch}</Fragment>
          })}
        </Fragment>,
      )
    }

    if (Array.isArray(children)) {
      return React.createElement(
        type,
        trimProps,
        <Fragment>
          {children.map((ch, idx) => (
            <Fragment key={idx}>{ch}</Fragment>
          ))}
        </Fragment>,
      )
    }

    return React.createElement(
      type,

      trimProps,
      children,
    )
  }

  return strictCreateElement
}

export const markdownOpts = {
  disableParsingRawHTML: false,
  forceBlock: true,
  overrides: {
    // Do not render by default.
    // The web and email renderers will register their corresponding implementations
    poll: () => <></>,
    Paywall: () => <></>,
    SubscribeNow: () => <></>,
    embed: () => <></>,
    iframe: () => <></>,

    // example style overrides
    img: (args: any) => <img {...args} style={{ maxWidth: '100%' }} />,
  },
} as const
