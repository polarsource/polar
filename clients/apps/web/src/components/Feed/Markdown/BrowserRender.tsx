import Markdown from 'markdown-to-jsx'
import dynamic from 'next/dynamic'
import { Skeleton } from 'polarkit/components/ui/skeleton'
import { createContext, useContext } from 'react'
import BrowserAd from './Ad/BrowserAd'
import Embed from './Embed/BrowserEmbed'
import Iframe from './Iframe/BrowserIframe'
import { ImageOverlay } from './Img/ImageOverlay'
import Paywall from './Paywall/Paywall'
import BrowserPoll from './Poll/BrowserPoll'
import Poll from './Poll/Poll'
import SubscribeNow from './SubscribeNow/SubscribeNow'
import {
  RenderArticle,
  markdownOpts,
  wrapStrictCreateElement,
} from './markdown'

const BrowserMermaid = dynamic(() => import('./Mermaid/BrowserMermaid'), {
  ssr: false,
  loading: () => (
    <Skeleton className="w-full p-8 font-mono text-sm">
      Initializing Mermaid Renderer...
    </Skeleton>
  ),
})

// Dynamically load the SyntaxHighlighter (heavily reduces bundle sizes)
//
// While loading (and SSR), render a placeholder without syntax highlighting
const SyntaxHighlighterContext = createContext('')
const BrowserSyntaxHighlighter = dynamic(
  () => import('./SyntaxHighlighter/BrowserSyntaxHighlighter'),
  {
    ssr: false,
    loading: () => <BrowserSyntaxHighlighterLoading />,
  },
)

const BrowserSyntaxHighlighterLoading = () => {
  const value = useContext(SyntaxHighlighterContext)
  return (
    <pre className="w-full">
      <code className="text-black dark:text-white">{value}</code>
    </pre>
  )
}

export const opts = {
  ...markdownOpts,

  overrides: {
    ...markdownOpts.overrides,

    // browser overrides
    poll: (args: any) => <Poll {...args} renderer={BrowserPoll} />,
    Paywall: (args: any) => <Paywall {...args} />,
    SubscribeNow: (args: any) => <SubscribeNow {...args} />,
    embed: (args: any) => <Embed {...args} />,
    iframe: (args: any) => <Iframe {...args} />,
    img: (args: any) => <ImageOverlay {...args} />,
    pre: (args: any) => {
      if (args.children.type === 'code') {
        const language = args.children.props.className?.replace('lang-', '')
        if (language === 'mermaid') {
          return (
            <BrowserMermaid graphDefinition={args.children.props.children} />
          )
        }
        return (
          <SyntaxHighlighterContext.Provider
            value={args.children.props.children}
          >
            <BrowserSyntaxHighlighter
              language={language}
              {...args.children.props}
            />
          </SyntaxHighlighterContext.Provider>
        )
      }
      return <></>
    },
    Ad: (args: any) => <BrowserAd {...args} />,
  },
} as const

export const previewOpts = {
  ...markdownOpts,
  overrides: {
    ...markdownOpts.overrides,

    poll: () => <></>,
    Paywall: () => <></>,
    SubscribeNow: () => <></>,
    embed: () => <></>,
    iframe: () => <></>,
    pre: () => <></>,
  },
} as const

const BrowserRender = ({
  article,
  showPaywalledContent,
  isSubscriber,
  paidArticlesBenefitName,
}: {
  article: RenderArticle
  showPaywalledContent: boolean
  isSubscriber: boolean
  paidArticlesBenefitName?: string
}) => {
  return (
    <Markdown
      options={{
        ...opts,
        createElement: wrapStrictCreateElement({
          article,
          showPaywalledContent,
          isSubscriber,
          paidArticlesBenefitName,
          extraAllowedCustomComponents: Object.keys(opts.overrides),
        }),
      }}
    >
      {article.body}
    </Markdown>
  )
}

export const AbbreviatedBrowserRender = ({
  article,
  showPaywalledContent,
}: {
  article: RenderArticle
  showPaywalledContent?: boolean
}) => {
  return (
    <Markdown
      options={{
        ...previewOpts,
        createElement: wrapStrictCreateElement({
          article,
          showPaywalledContent,
          isSubscriber: true, // Do not show <SubscribeNow /> in abbreviations
        }),
      }}
    >
      {article.body.substring(0, 500).split('\n').slice(0, 4).join('\n')}
    </Markdown>
  )
}

export default BrowserRender
