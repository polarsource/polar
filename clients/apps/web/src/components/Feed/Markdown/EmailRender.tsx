import Markdown from 'markdown-to-jsx'

import { Container } from '@react-email/components'
import Embed from './Embed/EmailEmbed'
import Iframe from './Iframe/EmailIframe'
import EmailMermaid from './Mermaid/EmailMermaid'
import Paywall, { EmailPaywall } from './Paywall/Paywall'
import Poll from './Poll/Poll'
import EmailSyntaxHighlighter from './SyntaxHighlighter/EmailSyntaxHighlighter'
import {
  RenderArticle,
  markdownOpts,
  wrapStrictCreateElement,
} from './markdown'

export const opts = {
  ...markdownOpts,
  overrides: {
    ...markdownOpts.overrides,

    // email overrides
    poll: (args: any) => <Poll {...args} renderer={EmailPoll} />,
    Paywall: (args: any) => <Paywall {...args} renderer={EmailPaywall} />,
    SubscribeNow: () => <></>, // do not render
    embed: (args: any) => <Embed {...args} />,
    iframe: (args: any) => <Iframe {...args} />,
    pre: (args: any) => {
      if (args.children.type === 'code') {
        const language = args.children.props.className?.replace('lang-', '')
        if (language === 'mermaid') {
          return <EmailMermaid graphDefinition={args.children.props.children} />
        }
        return (
          <EmailSyntaxHighlighter
            language={language}
            {...args.children.props}
          />
        )
      }
      return <></>
    },
  },
} as const

const EmailRender = (props: { article: RenderArticle }) => {
  return (
    <Markdown
      options={{
        ...opts,

        createElement: wrapStrictCreateElement({
          article: props.article,
        }),
      }}
    >
      {props.article.body}
    </Markdown>
  )
}

export default EmailRender

const EmailPoll = (props: { options: string[] }) => {
  return (
    <Container className="my-2 bg-green-300 p-8">
      <table className="w-full">
        {props.options.map((s) => (
          <tr key={s}>
            <td>{s}</td>
            <td className="text-right">123 votes (10%)</td>
          </tr>
        ))}
      </table>
    </Container>
  )
}
