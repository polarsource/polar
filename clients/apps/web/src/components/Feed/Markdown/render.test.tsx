import { article } from 'polarkit/testdata'

import '@testing-library/jest-dom'
import { render } from '@testing-library/react'
import Markdown from '@zegl/markdown-to-jsx'
import Paywall from './Paywall/Paywall'
import {
  RenderArticle,
  markdownOpts,
  wrapStrictCreateElement,
} from './markdown'
import { polarPost } from './testdata/polarPost'

// Currently unable to use any of the "real" renderers, as we can't import the syntax highlighter in jest (?)
const opts = {
  ...markdownOpts,
  overrides: {
    ...markdownOpts.overrides,
  },
} as const

const TestRenderer = (props: { article: RenderArticle }) => {
  return (
    <Markdown
      options={{
        ...opts,
        overrides: {
          ...opts.overrides,
          pre: () => <></>,
          Paywall: (args: any) => <Paywall {...args} />,
        },

        createElement: wrapStrictCreateElement({
          article: props.article,
        }),
      }}
    >
      {props.article.body}
    </Markdown>
  )
}

test('basic', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `
# h1

## h2
  
### h3

Hello **world**!

[Polar](https://polar.sh/)

<Paywall></Paywall>

This is a normal **block** of text [Polar](https://polar.sh/) with _various_ formatting.
And here it continues in the same block.

This is a different block.  
With a linebreak! (double whitespace)

> This is a quoute!
  `,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('double-whitespace', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: 'Normal **block** example\n\nWith-doublespace  \nWith-doublespace',
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('polar', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: polarPost,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('ads-embed', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `
        
<a href="https://www.youtube.com/watch?v=dQw4w9WgXcQ"><picture><img src="https://polar.sh/embed/ad?id=63f3a3cc-54ae-45e9-987a-7174364d234e" alt="Click me!!!!!" height="100" width="100" /></picture></a>
<a href="https://polar.sh/zegl"><picture><source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/embed/ad?id=94c1676c-db08-4489-b4b4-e25beadf2542&dark=1"><img src="https://polar.sh/embed/ad?id=94c1676c-db08-4489-b4b4-e25beadf2542" alt="Hello world!" height="100" width="100" /></picture></a>

        `,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('anchors', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `

<p id="top">
[Go to end](#end)
Here we go!
</p>


## End

<a href="#top">Go to top</a>
  
  `,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('XSS', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `

<p>
hello
</p>

<p onload="alert(1)">
what
</p>
    
<table>
    <thead>
        <tr><th>TH</th></tr>
    </thead>
    <tbody onload="alert(2)">
        <tr onload="alert(3)"><td onload="alert(4)">TD</td></tr>
        <tr><td><p onload="alert(1)"><img src="not-sanitized" onclick="xxx" /> <a href="javascript:alert(1)">clickme</a> <span onclick="xxx">clickmep</span></p></td><td onclick="x">TD</td></tr>
    </tbody>
</table>
  
  `,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('table', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `

| left  | center | right |
| ----- |:------:|----:|
| a | b | c |
| a | b | c |
| a | b | c |


<table>
    <thead>
        <tr><th>TH</th></tr>
    </thead>
    <tbody>
      <tr><td>TD</td></tr>
      <tr><td>TD</td></tr>
      <tr><td>TD</td></tr>
      <tr><td>TD</td></tr>
    </tbody>
</table>
        
  `,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('paywall', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `

prepaywall

<Paywall>here is the content! <strong>wow!</strong></Paywall>

<Paywall>plain content</Paywall>

<Paywall><img src="https://polar.sh/logo.png" /></Paywall>

<!-- Empty paywall contents! Should trigger upsell block! -->
<Paywall></Paywall>

postpaywall
  `,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('code', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `
 
code:::

\`\`\`go
func main() {
}
\`\`\`

`,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('blockquote', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `
 

And code blocks:

<blockquote>
i am a block
</blockquote>

<blockquote>
multi

line

block
</blockquote>

`,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('footnote', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `
Here's a simple footnote,[^1] and here's another one[^ref-abc].

[^1]: This is the **first** footnote.
[^ref-abc]: This is \`another\` note!

`,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('references', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `
In a hole in the ground there lived a hobbit. Not a nasty, dirty, wet hole, filled with the ends
of worms and an oozy smell, nor yet a dry, bare, sandy hole with nothing in it to sit down on or to
eat: it was a [hobbit-hole][1], and that means comfort.

[1]: <https://en.wikipedia.org/wiki/Hobbit#Lifestyle> "Hobbit lifestyles"
`,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('link-brackets-code', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `E.g [\`[pyclass]\`](https://pyo3.rs/v0.20.3/class#defining-a-new-class).`,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})

test('link-brackets', () => {
  const { container } = render(
    <TestRenderer
      article={{
        ...article,
        body: `WOW! [[wow]](https://polar.sh) :-)`,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})
