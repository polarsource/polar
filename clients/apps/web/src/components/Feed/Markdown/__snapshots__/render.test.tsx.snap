// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ads-embed 1`] = `
<div>
  <div>
    <a
      href="https://www.youtube.com/watch?v=dQw4w9WgXcQ"
    >
      <picture>
        <img
          alt="Click me!!!!!"
          height="100"
          src="https://polar.sh/embed/ad?id=63f3a3cc-54ae-45e9-987a-7174364d234e"
          style="max-width: 100%;"
          width="100"
        />
      </picture>
    </a>
    <a
      href="https://polar.sh/zegl"
    >
      <picture>
        <source
          media="(prefers-color-scheme: dark)"
          srcset="https://polar.sh/embed/ad?id=94c1676c-db08-4489-b4b4-e25beadf2542&dark=1"
        />
        <img
          alt="Hello world!"
          height="100"
          src="https://polar.sh/embed/ad?id=94c1676c-db08-4489-b4b4-e25beadf2542"
          style="max-width: 100%;"
          width="100"
        />
      </picture>
    </a>
  </div>
</div>
`;

exports[`anchors 1`] = `
<div>
  <div>
    <p
      id="top"
    >
      <a
        href="#end"
      >
        Go to end
      </a>
      
Here we go!

    </p>
    <h2
      id="end"
    >
      End
    </h2>
    <a
      href="#top"
    >
      Go to top
    </a>
  </div>
</div>
`;

exports[`basic 1`] = `
<div>
  <div>
    <h1
      id="h1"
    >
      h1
    </h1>
    <h2
      id="h2"
    >
      h2
    </h2>
    <h3
      id="h3"
    >
      h3
    </h3>
    <p>
        Hello 
      <strong>
        world
      </strong>
      !
    </p>
    <p>
        
      <a
        href="https://polar.sh/"
      >
        Polar
      </a>
    </p>
    <div
      class="my-4 flex flex-col items-center rounded-3xl bg-gray-100 px-8 py-4 text-center dark:bg-polar-700"
    >
      <p>
        This section is for premium subscribers only. Subscribe to
         
        <a
          class="text-blue-500 hover:text-blue-400 dark:text-blue-400 dark:hover:text-blue-300"
          href="/pydantic/subscriptions"
        >
          Pydantic
        </a>
         
        to get access to it.
      </p>
    </div>
    <p>
        This is a normal 
      <strong>
        block
      </strong>
       of text 
      <a
        href="https://polar.sh/"
      >
        Polar
      </a>
       with 
      <em>
        various
      </em>
       formatting.
  And here it continues in the same block.
    </p>
    <p>
        This is a different block.
      <br />
        With a linebreak! (double whitespace)
    </p>
    <blockquote>
      <p>
        This is a quoute!
      </p>
    </blockquote>
  </div>
</div>
`;

exports[`double-whitespace 1`] = `
<div>
  <div>
    <p>
      Normal 
      <strong>
        block
      </strong>
       example
    </p>
    <p>
      With-doublespace
      <br />
      With-doublespace
    </p>
  </div>
</div>
`;

exports[`polar 1`] = `
<div>
  <div>
    <p>
      <img
        alt="image.png"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-fCQqIZt3zubCpaMQzy6ZONZAOG6QDq.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      Today, we're thrilled to announce & launch the next chapter of Polar.
    </p>
    <p>
      A creator platform for developers and the open source ecosystem â€“Â built open source (
      <a
        href="https://github.com/polarsource/polar"
      >
        fork us on GitHub
      </a>
      ).
    </p>
    <p>
      Offering you â€“ as a developer â€“Â  a platform on top of your GitHub repositories to:
    </p>
    <ol>
      <li>
        Build, own & reach your audience through free- and premium posts and newsletters.
      </li>
      <li>
        Offer subscriptions of value-add benefits designed for our ecosystem & built-in to Polar. We'll also handle value-add taxes so you don't have to.
      </li>
      <li>
        Ability to integrate it all on your own docs, sites or services using our API & SDK.
      </li>
    </ol>
    <p>
      We're also excited to share our updated pricing (it's better for you) and that our team is going to subscribe to open source developers on Polar for $30K in 2024 ðŸŽ‰
    </p>
    <p>
      Let's dive in.
    </p>
    <h3
      id="subscriptions--benefits"
    >
      Subscriptions & Benefits
    </h3>
    <p>
      <img
        alt="Screenshot 2024-01-09 at 10.44.55.png"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/Screenshot%202024-01-09%20at%2010.44.55-uy562dyst8J1Hk4QScE5i6DE3Aa2jA.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      You can now setup subscriptions with Polar ðŸš€ They're beautiful & come with some incredibly powerful features beyond what you'd expect.
    </p>
    <p>
      <strong>
        Reach your entire audience with free tier(s)
      </strong>
    </p>
    <p>
      You automatically have a built-in - free - tier offering newsletter access to your Public Posts on Polar. Of course, you can edit it to add even more benefits too.
    </p>
    <p>
      Enabling your entire audience to subscribe with email-only at first and for you to have one platform to reach them all.
    </p>
    <p>
      <strong>
        Paid tiers & managed value-add taxes
      </strong>
    </p>
    <p>
      Of course, you can offer paid tiers too that backers can upgrade to seamlessly over time.
    </p>
    <p>
      Many developers shy away from offering paid services & benefits due to the value-add tax implications. Calculating, capturing & remitting value-add taxes internationally is a crushing overhead for individual developers.
    </p>
    <p>
      That's why we're incredibly proud to be the merchant of record and handle value-add taxes so you don't have to. Read more about the details 
      <a
        href="https://docs.polar.sh/payment/"
      >
        here
      </a>
      .
    </p>
    <p>
      <strong>
        Unparalleled business offerings
      </strong>
    </p>
    <p>
      Open source developers tend to have sponsorship tiers for individual users vs. businesses, but they're all treated the same. Not with Polar.
    </p>
    <p>
      Polar has the concept of subscription audiences: Individual vs. Business. Separating them publicly for clarity, but also ensuring benefits are granted to all individual team members of the organization that subscribes ðŸ¤¯
    </p>
    <p>
      Offering unparalleled opportunities and value to subscriptions offered via Polar to businesses.
    </p>
    <p>
      <strong>
        Powerful, built-in, benefits
      </strong>
    </p>
    <p>
      You can setup custom benefits with Polar (free text) for maximum flexibility. However, we're going to invest heavily in automating typical benefits offered within our ecosystem. Making it seamless for you to manage & delightful for your backers to use.
    </p>
    <p>
      From Discord invites (roles/tier) to automating how businesses can setup their logotype & description once and it automatically propagating across your README (PR/SVG) to sites/docs (SDK). Or how about early access to new features, i.e invite to private repositories? Sponsored posts? Just to mention a few ideas in our pipeline.
    </p>
    <p>
      Our first built-in benefit is Posts & Newsletters. Enabling a vital communication channel to an audience you own & can reach fully - regardless of which benefits you offer today or tomorrow.
    </p>
    <h3
      id="posts--newsletters"
    >
      Posts & Newsletters
    </h3>
    <p>
      <img
        alt="image.png"
        src="https://7vk6rcnylug0u6hg.public.blob.vercel-storage.com/image-QKge9Wh9vpdhZRIPLnsMWFesg8SPEY.png"
        style="max-width: 100%;"
      />
    </p>
    <p>
      It's a bit meta, but you're reading a Polar post right now ðŸ˜Ž It's written in markdown, I previewed it easily both on the web & email and I can share it publicly or to paid subscribers.
    </p>
    <p>
      Or how about a hybrid? This is a public post, but the below is a section reserved for paid subscribers. You see that by the upgrade module below (don't worry - I only hid "1337" in there :-)
    </p>
    <p>
      1337
    </p>
    <p>
      We love to write & share code too, so of course we support syntax highlighting in Polar Posts.
    </p>
    <p>
      I can embed Youtube videos, images, links or even GitHub issues you can fund using Polar issue funding & rewards.
    </p>
    <p>
      We've battled with the email gods to make all of this a delightful experience over email too. I can easily choose to share this post over the web only or also email it to all my subscribers (newsletter).
    </p>
    <p>
      Now might be a good time to promote subscribing to our own Polar Posts in the future. So let me embed that nudge easily below.
    </p>
    <p>
      We're working on polls and other embeddables too. We can't wait to hear what you think & how we can improve posts for your initiatives!
    </p>
    <h3
      id="api--sdk"
    >
      API & SDK
    </h3>
    <p>
      You get a personal page on Polar that you can easily promote Posts, Subscriptions & Issue Funding from Day 1: e.g 
      <code>
        polar.sh/polarsource
      </code>
      .
    </p>
    <p>
      We don't think it makes any sense to offer customization abilities to our audience; setting up custom color hex codes, templates etc like other creator platforms. Instead, we're building everything on our public API & offering SDKs (JS first - beta). Since we believe you'd rather integrate it directly on your site, docs and/or services.
    </p>
    <p>
      Our goal is that the developer page, e.g 
      <code>
        polar.sh/polarsource
      </code>
      , should be a living example of an API/SDK integration that you can cherry-pick components from. We'd love to hear from you and your use case and work together to ensure we support it - 
      <a
        href="https://discord.gg/AVgbMhTtBq"
      >
        join our Discord
      </a>
       to chat about it.
    </p>
    <p>
      In the meantime, don't hesitate to checkout our:
    </p>
    <ul>
      <li>
        <a
          href="https://api.polar.sh/redoc"
        >
          Public Polar API
        </a>
      </li>
      <li>
        <a
          href="https://github.com/polarsource/polar/releases/tag/jssdk%2Fv0.2.0"
        >
          JS SDK
        </a>
      </li>
    </ul>
    <p>
      <em>
        It's in early development and we'd love your feedback. We are mindful to maintain backward compatibility, but cannot guarantee it at this stage.
      </em>
    </p>
    <h3
      id="pricing"
    >
      Pricing
    </h3>
    <p>
      We're incredibly excited to share that we're updating our pricing with this launch. In order to drive even more capital to open source developers.
    </p>
    <p>
      Moving forward, our pricing is:
    </p>
    <ul>
      <li>
        No fixed costs
      </li>
      <li>
        5% + payment & payout fees (Stripe) instead of our former 10% + payment & payout fees.
      </li>
      <li>
        <em>
          We'll also cover Stripe fees from our 5% until March 31st, 2024
        </em>
        .
      </li>
    </ul>
    <p>
      Let's go ðŸš€
    </p>
    <h3
      id="30k-in-2024-from-team-polar"
    >
      $30K in 2024 from Team Polar
    </h3>
    <p>
      Finally, we're happy to share that we're going to invest $30K of our own to subscribe to open source developers throughout 2024. So don't hesitate to ping @polar_sh on Twitter with your posts to share it with us and we might subscribe :-)
    </p>
    <p>
      How is this funded?
    </p>
    <ol>
      <li>
        Birk lost it in 2023 and pledged $30K personally and has $20K left to spend.
      </li>
      <li>
        Polar supports businesses to setup open source allowances for team members (invite-only for now). Polar has an employee benefit of each engineer getting $200/month. We're 4 people so that's $800/month or $9,600/year. 
        <em>
          We're rounding up the rest :-)
        </em>
      </li>
    </ol>
    <hr />
    <p>
      Phew. That's all. This is our biggest update since launching Polar and a big step towards our long-term vision of building a platform for developers to get more funding to even building independent businesses around their initiatives.
    </p>
    <p>
      As always, don't hesitate to share your feedback, thoughts or requests on 
      <a
        href="https://github.com/polarsource/polar"
      >
        GitHub
      </a>
       or chat with us directly on 
      <a
        href="https://discord.gg/AVgbMhTtBq"
      >
        Discord
      </a>
      . We're here to serve fellow developers!
    </p>
    <p>
      All our best,
      <br />
      Birk, Gustav, Francois & Emil @ Polar
    </p>
  </div>
</div>
`;
